designs/variables-for-userdata.md

```diff
@@ -0,0 +1,206 @@
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: default
+spec:
+  amiFamily: Custom
+  # New field for non-drifting variables
+  userDataVariables:
+    AUTH_TOKEN: "rotated-secret-token-value"
+  userData: | # an example of a minimal ignition config userData
+  '{
+    "ignition": {
+      "config": {
+        "merge": [
+          {
+            "httpHeaders": [
+              {
+                "name": "Authorization",
+                "value": "Bearer AUTH_TOKEN"
+              }
+            ],
+            "source": "https://bootstrap.internal/config"
+          }
+
+        ]
+      }
+    }
+  }'
+```
+
+#### Implementation Details
+
+The code's AMIFamily resolver would need to:
+
+1. Parse the `userData` template and identify variable tokens by exact name (e.g., AUTH_TOKEN).
+2. Substitute variables with their current values from `userDataVariables` before computing the launch template at runtime.
+
+The code's drift calculator will need to:
+
+1. When calculating the `EC2NodeClass` `Hash()` function, which is used in drift detection, ignore the `userDataVariables` field by tagging the field with `hash:"ignore"`.
+
+#### Advantages
+
+* Simple to use
+* Flexible to work with any `userData` format (shell scripts, cloud-config, etc.) and any `AMIFamily`.
+* Compliant with current drift static field ignoring mechanisms.
+
+#### Disadvantages
+
+* Only supports simple string substitution.
+* Potentially sensitive data stored in the cluster would still be stored in plaintext in the `EC2NodeClass` spec.
+
+### Option 2: Reference Kubernetes Secrets
+
+This option introduces a new field, `userDataSecretRefs`, to the `EC2NodeClass` spec. This field is a list of references to Kubernetes Secrets, which will be used to populate the `userData` template variables. This is a more secure way to inject dynamic data into the `userData` field, as the secrets are not stored in the `EC2NodeClass` spec in plaintext. This also solves the drift issue, as the secret's underlying data can change, but that is not visible to the drift detection mechanism.
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: default
+spec:
+  userDataSecretRefs:
+    - name: bootstrap-token-secret # Name of the Secret
+      namespace: some-namespace    # Namespace where of the Secret
+      key: token                  # Key within the Secret
+      variable: AUTH_TOKEN        # Variable name to use in the template
+  userData: |
+    #!/bin/bash
+    curl --header "Authorization: Bearer AUTH_TOKEN" "https://bootstrap.internal/config"
+```
+
+#### Implementation Details
+
+* The Karpenter controller would require RBAC permissions to get and watch Secrets.
```

Comment:
This seems problematic as well since you'd need Karpenter to potentially access all secrets in all namespaces.